tensorflow_probability/python/sts/trigonometric.py

```diff
@@ -0,0 +1,225 @@
+      self._period = period
+      self._selected_frequencies = selected_frequencies
+
+      super(TrigonometricStateSpaceModel, self).__init__(
+          num_timesteps=num_timesteps,
+          transition_matrix=transition_matrix,
+          transition_noise=tfd.MultivariateNormalDiag(
+              scale_diag=(drift_scale[..., tf.newaxis] *
+                          tf.ones([2 * num_frequencies], dtype=dtype)),
+              name='transition_noise'),
+          observation_matrix=observation_matrix,
+          observation_noise=tfd.MultivariateNormalDiag(
+              scale_diag=observation_noise_scale[..., tf.newaxis],
+              name='observation_noise'),
+          initial_state_prior=initial_state_prior,
+          initial_step=initial_step,
+          allow_nan_stats=allow_nan_stats,
+          validate_args=validate_args,
+          name=name)
+
+  @property
+  def drift_scale(self):
+    """Standard deviation of the drift in the cyclic effects."""
+    return self._drift_scale
+
+  @property
+  def observation_noise_scale(self):
+    """Standard deviation of the observation noise."""
+    return self._observation_noise_scale
+
+  @property
+  def period(self):
+    """The seasonal period."""
+    return self._period
+
+  @property
+  def selected_frequencies(self):
+    """Chosen frequency indices to include."""
+    return self._selected_frequencies
+
+
+def build_trigonometric_transition_matrix(period, selected_frequencies, dtype):
+  """Build the transition matrix for a TrigonometricStateSpaceModel"""
+
+  two_pi = tf.constant(2. * np.pi, dtype=dtype)
+  frequencies = two_pi * selected_frequencies / period
+  num_frequencies = len(selected_frequencies)
+
+  sin_frequencies = tf.sin(frequencies)
+  cos_frequencies = tf.cos(frequencies)
+
+  trigonometric_values = tf.stack(
+      [cos_frequencies, sin_frequencies, -sin_frequencies, cos_frequencies],
+      axis=-1)
+
+  transition_matrix = tf.linalg.LinearOperatorBlockDiag(
+      [
+          tf.linalg.LinearOperatorFullMatrix(
+              matrix=tf.reshape(trigonometric_values[i], [2, 2]),
+              is_square=True
+          )
+          for i in range(num_frequencies)
+      ]
+  )
+
+  return transition_matrix
+
+
+class Trigonometric(StructuralTimeSeries):
+  """Formal representation of a trigonometric seasonal effects model."""
+
+  def __init__(self,
+               period,
+               selected_frequencies,
+               drift_scale_prior=None,
+               initial_effect_prior=None,
+               observed_time_series=None,
+               name=None):
+    """Specify a trigonometric seasonal effects model."""
+
+    with tf.compat.v1.name_scope(
+        name, 'Trigonometric', values=[observed_time_series]) as name:
+
+      _, observed_stddev, observed_initial = (
+          sts_util.empirical_statistics(observed_time_series)
+          if observed_time_series is not None else (0., 1., 0.))
+
+      latent_size = 2 * len(selected_frequencies)
```

Comment:
Cool! I wanted to do something like this but I didn't know how to. Hopefully this is addressed in ed8a438aa2e4db30d6b7ac5bc3811304b25f9932. I also tried to address the comment below about the priors in this commit as I thought it would be easier but looking back I'm actually not sure what I was thinking - sorry about that.